src/experimental/tools/videoThumbnailLoader/thumbnail_loader.ts

```diff
@@ -0,0 +1,307 @@
+        }
+
+        const thumbnails: IThumbnailInfo[] | undefined =
+          this._thumbnailTrack.thumbnailInfos
+            .filter((t) => {
+              const thumbnailDuration = t.duration;
+              const range = { start: payload.time - thumbnailDuration,
+                              end: payload.time + thumbnailDuration };
+              const tRange = { start: t.start, end: t.start + t.duration };
+              const rangesAreOverlapping = areRangesOverlapping(range, tRange);
+              let hasBuffered = false;
+              const { buffered } = videoElement;
+              for (let i = 0; i < buffered.length; i++) {
+                if (buffered.start(i) <= tRange.start &&
+                    buffered.end(i) >= tRange.end)
+                {
+                  hasBuffered = true;
+                }
+              }
+              return rangesAreOverlapping && !hasBuffered;
+            });
+
+      if (thumbnails.length === 0) {
+        throw new Error(
+          "VideoThumbnailLoaderError: Couldn't find thumbnail.");
+      }
+
+      log.debug("VTL: Found thumbnails for time", payload.time, thumbnails);
+
+      return { thumbnails, payload };
+    }),
+    distinctUntilChanged((a, b) => {
+      if (a.thumbnails.length !== b.thumbnails.length) {
+        return false;
+      }
+      for (let i = 0; i < a.thumbnails.length; i++) {
+        if (a.thumbnails[i].start !== b.thumbnails[i].start ||
+            a.thumbnails[i].duration !== b.thumbnails[i].duration ||
+            a.thumbnails[i].mediaURL !== b.thumbnails[i].mediaURL) {
+          return false;
+        }
+      }
+      return true;
+    }),
+    switchMap(({ thumbnails, payload: { time, resolve, reject } }) => {
+      return videoSourceInfos$.pipe(
+        mergeMap((videoSourceBuffer) => {
+          return videoSourceBuffer.removeBuffer(time - 5, time + 5).pipe(
+            mergeMap(() => {
+              if (!this._thumbnailTrack) {
+                throw new Error(
+                  "VideoThumbnailLoaderError: No thumbnail track given.");
+              }
+
+              log.debug("VTL: Removed buffer before appending segments.", time);
+
+              return getSegmentsData(thumbnails, videoElement).pipe(
+                mergeMap((data) => {
+                  if (data) {
+                    const appendBuffer$ = videoSourceBuffer
+                      .appendSegment({
+                        chunk: data,
+                        initSegment: null,
+                        codec: this._thumbnailTrack.codec });
+                    return appendBuffer$.pipe(
+                      tap(() => {
+                        log.debug("VTL: Appended segment.", data, time);
+                        this._thumbnailVideoElement.currentTime = time;
+                      })
+                    );
+                  }
+                  return observableOf(null);
+                }),
+                finalize(() => {
+                  resolve();
```

Comment:
I don't think resolving in a `finalize` is a good idea:
  1. the time at which it is called is not very clear. It should be reserved for clean up
  2. It will be called at any time: in error, canceled or success.